File: sample_apps/js/utils/timestream-dependency-helper.js
import {
    S3Client,
    CreateBucketCommand,
    DeleteBucketCommand,
    ListObjectsCommand,
    DeleteObjectsCommand, GetObjectCommand
} from "@aws-sdk/client-s3";
import {
    STSClient,
    GetCallerIdentityCommand
} from "@aws-sdk/client-sts";
import zlib from 'zlib';
import { Readable } from 'stream';
import csvParser from "csv-parser";

export class TimestreamDependencyHelper {
    constructor(region) {
        this.s3Client = new S3Client({region: region});
        this.stsClient = new STSClient();
    }

    async getAccount() {
        const params = new GetCallerIdentityCommand({});
        return await this.stsClient.send(params).then(
            (data) => {
                return data.Account;
            },
            (err) => {
                console.log("Error while getting the account.", err);
                throw err;
            }
        )
    }

    async createS3Bucket(bucketName) {
        const params = new CreateBucketCommand({
            Bucket: bucketName
        })
        return await this.s3Client.send(params).then(
            (data) => {
                console.log(`Bucket [%s] created`, bucketName);
            },
            (err) => {
                console.log("Error while creating the bucket.", err);
            }
        )
    }

    async deleteS3Bucket(bucketName) {
        console.log("Deleting S3 Bucket");
        const { Contents } = await this.listS3Objects(bucketName);

        if (Contents && Contents.length > 0) {
            await this.deleteS3Objects(bucketName, Contents);
        }
        const params = new DeleteBucketCommand({
            Bucket: bucketName
        })
        return await this.s3Client.send(params).then(
            (data) => {
                console.log("Bucket [%s] deleted Successfully", bucketName);
            },
            (err) => {
                console.log("Deletion of S3 error report bucket failed: ", err);
        })
    }

    async deleteS3Objects(bucketName, contents){
        const params = new DeleteObjectsCommand({
            Bucket: bucketName,
            Delete: {
                Objects: contents.map(({Key}) => ({Key}))
            }
        })
        return await this.s3Client.send(params).then((data) => {
            console.log("Objects deleted!");
        }, (err) => {
            console.error("Error while deleting the objects", err);
        });
    }

    async getS3Object(uri) {
        const {bucketName, key} = this.getBucketAndKey(uri);
        const params = new GetObjectCommand({
            Bucket: bucketName,
            Key: key
        })
        const response = await this.s3Client.send(params);
        return await response.Body.transformToString();
    }

    async getS3CSVObject(uri, headers) {
        const {bucketName, key} = this.getBucketAndKey(uri);
        const params = new GetObjectCommand({
            Bucket: bucketName,
            Key: key
        });
        const response = await this.s3Client.send(params);
        try {
            let results = [];
            return await new Promise((resolve, reject) => {
                const stream = Readable.from(response.Body);
                stream.pipe(zlib.createGunzip()).pipe(
                    csvParser({separator: ',', headers: false})
                ).on("data", (chunk) => { results.push(chunk);})
                .on("end", () => { resolve(results);});
        });
        } catch (e) {
            console.error("Error while reading the file", e);
        }
    }

    async listS3Objects(bucketName) {
        const params = new ListObjectsCommand({
            Bucket: bucketName
        })
        return await this.s3Client.send(params).then(
            (data) => {
                console.log("Objects listed!")
                return data;
            },
            (err) => {
                console.error("Unable to list Objects");
            }
        )
    }

    getBucketAndKey(uri) {
        const [bucketName] = uri.replace("s3://", "").split("/", 1);
        const key = uri.replace("s3://", "").split('/').slice(1).join('/');
        return {bucketName, key};
    }

     generateRandomStringWithSize(size) {
        const chars = "abcdefghijklmnopqrstuvwxyz0123456789"
        let result = "";
        for (let i = 0; i < size; i++) {
            result += chars.charAt(Math.floor(Math.random() * chars.length));
        }
        return result;
    }

}

File: sample_apps/python/live_analytics/basic_example/utils/TimestreamDependencyHelper.py
import boto3
import json


class TimestreamDependencyHelper:

    def __init__(self, region):
        self.region = region
        self.s3_client = boto3.client('s3', region_name=region)
        self.s3_resource = boto3.resource('s3')
        self.sts_client = boto3.client("sts")

    def get_account(self):
        return self.sts_client.get_caller_identity()["Account"]

    def create_s3_bucket(self, bucket_name):
        print("\nCreating S3 bucket " + bucket_name)
        try:
            if self.region == 'us-east-1':
                self.s3_client.create_bucket(Bucket=bucket_name)
            else:
                self.s3_client.create_bucket(Bucket=bucket_name,
                                             CreateBucketConfiguration={'LocationConstraint': self.region})
            print("Successfully created S3 bucket: ", bucket_name)
            return bucket_name

        except self.s3_client.exceptions.BucketAlreadyOwnedByYou:
            print("Skipping Bucket Creation. Reason: Bucket " + bucket_name + " already exists.")
        except Exception as err:
            print("Failed creating S3 bucket: ", err)
            raise err

    def delete_s3_bucket(self, bucket_arn):
        print("Deleting S3 bucket")
        try:
            bucket = self.s3_resource.Bucket(bucket_arn)
            bucket.objects.all().delete()
            bucket.delete()
            print("Successfully deleted S3 bucket: ", bucket_arn)
        except Exception as err:
            # Not raising an exception here as we want other cleanup to continue
            print("Failed to delete S3 bucket: ", err)

    def list_s3_objects(self, bucket_name, prefix):
        print("\nListing object in S3 bucket")
        try:
            bucket = self.s3_resource.Bucket(bucket_name)
            print("Successfully listed objects in S3 bucket: ", bucket_name)
            return bucket.objects.filter(Prefix=prefix)
        except Exception as err:
            # Not raising an exception here as we want other cleanup to continue
            print("Failed to list objects in S3 bucket: ", err)
    
    def get_object(self, uri):
        try:
            bucket, key = uri.replace("s3://", "").split("/", 1)
            response = self.s3_client.get_object(Bucket=bucket, Key=key)
            return response['Body']
        except Exception as err:
            print("Failed to get the object for URI:", uri)
            raise err
